File: services/radio_api.py
import requests

SERVERS = [
    "https://de1.api.radio-browser.info",
    "https://nl1.api.radio-browser.info",
]
TIMEOUT = 10
USER_AGENT = "AirStream/1.0"


def _get(path, params=None):
    headers = {"User-Agent": USER_AGENT}
    for server in SERVERS:
        try:
            resp = requests.get(
                f"{server}{path}",
                params=params,
                headers=headers,
                timeout=TIMEOUT,
            )
            resp.raise_for_status()
            return resp.json()
        except (requests.RequestException, ValueError):
            continue
    return []


def fetch_top_stations(limit=100):
    return _get(f"/json/stations/topclick/{limit}", {"hidebroken": "true"})


def search_stations(name="", countrycode="", tag="", limit=100):
    params = {
        "order": "clickcount",
        "reverse": "true",
        "limit": str(limit),
        "hidebroken": "true",
    }
    if name:
        params["name"] = name
    if countrycode:
        params["countrycode"] = countrycode
    if tag:
        params["tag"] = tag
    return _get("/json/stations/search", params)


def fetch_countries(limit=50):
    return _get("/json/countries", {
        "order": "stationcount",
        "reverse": "true",
        "limit": str(limit),
    })


def fetch_tags(limit=50):
    return _get("/json/tags", {
        "order": "stationcount",
        "reverse": "true",
        "limit": str(limit),
    })


def fetch_stations_by_uuid(uuids):
    if not uuids:
        return []
    return _get("/json/stations/byuuid", {"uuids": ",".join(uuids)})

File: mobile/lib/services/radio_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/station.dart';

class RadioApi {
  static const _servers = [
    'https://de1.api.radio-browser.info',
    'https://nl1.api.radio-browser.info',
  ];
  static const _timeout = Duration(seconds: 10);

  static Future<List<dynamic>> _get(String path,
      {Map<String, String>? params}) async {
    for (final server in _servers) {
      try {
        final uri = Uri.parse('$server$path').replace(queryParameters: params);
        final response = await http
            .get(uri, headers: {'User-Agent': 'AirStream/1.0'}).timeout(
                _timeout);
        if (response.statusCode == 200) {
          return jsonDecode(response.body) as List<dynamic>;
        }
      } catch (_) {
        continue;
      }
    }
    return [];
  }

  static Future<List<Station>> fetchTopStations({int limit = 100}) async {
    final data =
        await _get('/json/stations/topclick/$limit', params: {'hidebroken': 'true'});
    return data.map((j) => Station.fromJson(j)).toList();
  }

  static Future<List<Station>> searchStations({
    String name = '',
    String countrycode = '',
    String tag = '',
    int limit = 100,
  }) async {
    final params = <String, String>{
      'order': 'clickcount',
      'reverse': 'true',
      'limit': '$limit',
      'hidebroken': 'true',
    };
    if (name.isNotEmpty) params['name'] = name;
    if (countrycode.isNotEmpty) params['countrycode'] = countrycode;
    if (tag.isNotEmpty) params['tag'] = tag;

    final data = await _get('/json/stations/search', params: params);
    return data.map((j) => Station.fromJson(j)).toList();
  }

  static Future<List<Map<String, dynamic>>> fetchCountries(
      {int limit = 50}) async {
    final data = await _get('/json/countries', params: {
      'order': 'stationcount',
      'reverse': 'true',
      'limit': '$limit',
    });
    return data.cast<Map<String, dynamic>>();
  }

  static Future<List<Map<String, dynamic>>> fetchTags(
      {int limit = 50}) async {
    final data = await _get('/json/tags', params: {
      'order': 'stationcount',
      'reverse': 'true',
      'limit': '$limit',
    });
    return data.cast<Map<String, dynamic>>();
  }
}
